src/app/shared/spaces.service.ts

```diff
@@ -1,72 +1,84 @@
-      contexts.current
-        .map(val => val.space),
-      this.broadcaster.on<Space>('spaceUpdated'));
-    // Create the recent context list
+      this.contexts.current.map(val => val.space),
+      this.broadcaster.on<Space>('spaceUpdated')
+    );
+  }
+
+  private initRecent(): void {
+    // load existing recent spaces from profile.store.recentSpaces
+    this.loadRecent().subscribe(spaces => {
+      this._recent.next(spaces);
+    });
 
-    this._recent = Observable.merge(
-      this.broadcaster.on<Space>('spaceChanged'),
-      this.addRecent)
-      // Map from the context being added to an array of recent contexts
-      // The scan operator allows us to access the list of recent contexts and add ours
-      .scan((recent, s) => {
-        // First, check if this context is already in the list
-        // If it is, remove it, so we don't get duplicates
-        for (let i = recent.length - 1; i >= 0; i--) {
-          if (recent[i].id === s.id) {
-            recent.splice(i, 1);
+    // update _recent with newly seen space when spaceChanged is emitted
+    this.broadcaster.on<Space>('spaceChanged')
+      .flatMap((changedSpace: Space) => {
+        return this._recent.first().map((recentSpaces: Space[]) => {
+          let index: number = this.findRecentIndexById(recentSpaces, changedSpace.id);
+          // if changedSpace exists in _recent
+          if (index !== -1) {
+            // move the space to the front of the list if it is already in _recent
+            if (recentSpaces[0].id !== changedSpace.id) {
+              recentSpaces.splice(index, 1);
+              recentSpaces.unshift(changedSpace);
+              this.saveRecent(recentSpaces);
+              this._recent.next(recentSpaces);
+            }
+            // otherwise, no operation required; changedSpace is already index 0
+          // if changedSpace is new to the recent spaces
+          } else {
+            recentSpaces.unshift(changedSpace);
+            // trim _recent if required to ensure it is length =< 8
+            if (recentSpaces.length > SpacesService.RECENT_SPACE_LENGTH) {
+              recentSpaces.pop();
+            }
+            this.saveRecent(recentSpaces);
+            this._recent.next(recentSpaces);
           }
-        }
-        // Then add this context to the top of the list
-        recent.unshift(s);
-        return recent;
-        // The final value to scan is the initial value, used when the app starts
-      },
-      [])
-      // Finally save the list of recent contexts
-      .do(val => {
-        // Truncate the number of recent contexts to the correct length
-        if (val.length > SpacesService.RECENT_SPACE_LENGTH) {
-          val.splice(
-            SpacesService.RECENT_SPACE_LENGTH,
-            val.length - SpacesService.RECENT_SPACE_LENGTH
-          );
-        }
+          return recentSpaces;
+        });
       })
-      .do(val => {
-        this.saveRecent(val);
+      .subscribe(() => {});
```

Comment:
This empty subscriber is pretty weird. Could you not move the `this._recent.next()` part of the `flatMap` into this subscriber? It is also weird that you have that side-effect within your mapping.

Reply:
That's fair, my mistake I had interpreted Jie's review comment [0] as a request to change rather than a question when I made that change. I had the saving logic in the subscription and used a boolean flag to only save if the list was manipulated in the flatMap, I'll go back to that.

[0] https://github.com/fabric8-ui/fabric8-ui/pull/3211#discussion_r214964039